actionpack/lib/action_dispatch/http/url.rb

```diff
@@ -3,179 +3,15 @@
-            if anchor
-              path << "##{Journey::Router::Utils.escape_fragment(anchor.to_param)}"
-            end
-          end
-
-          def extract_domain_from(host, tld_length)
-            host.split(".").last(1 + tld_length).join(".")
-          end
-
-          def extract_subdomains_from(host, tld_length)
-            parts = host.split(".")
-            parts[0..-(tld_length + 2)]
-          end
-
-          def build_host_url(host, port, protocol, options, path)
-            if match = host.match(HOST_REGEXP)
-              protocol ||= match[1] unless protocol == false
-              host       = match[2]
-              port       = match[3] unless options.key? :port
-            end
-
-            protocol = normalize_protocol protocol
-            host     = normalize_host(host, options)
-
-            result = protocol.dup
-
-            if options[:user] && options[:password]
-              result << "#{Rack::Utils.escape(options[:user])}:#{Rack::Utils.escape(options[:password])}@"
-            end
-
-            result << host
-            normalize_port(port, protocol) { |normalized_port|
-              result << ":#{normalized_port}"
-            }
-
-            result.concat path
-          end
-
-          def named_host?(host)
-            !IP_HOST_REGEXP.match?(host)
-          end
-
-          def normalize_protocol(protocol)
-            case protocol
-            when nil
-              secure_protocol ? "https://" : "http://"
-            when false, "//"
-              "//"
-            when PROTOCOL_REGEXP
-              "#{$1}://"
-            else
-              raise ArgumentError, "Invalid :protocol option: #{protocol.inspect}"
-            end
-          end
-
-          def normalize_host(_host, options)
-            return _host unless named_host?(_host)
-
-            tld_length = options[:tld_length] || @@tld_length
-            subdomain  = options.fetch :subdomain, true
-            domain     = options[:domain]
-
-            host = +""
-            if subdomain == true
-              return _host if domain.nil?
-
-              host << extract_subdomains_from(_host, tld_length).join(".")
-            elsif subdomain
-              host << subdomain.to_param
-            end
-            host << "." unless host.empty?
-            host << (domain || extract_domain_from(_host, tld_length))
-            host
-          end
-
-          def normalize_port(port, protocol)
-            return unless port
-
-            case protocol
-            when "//" then yield port
-            when "https://"
-              yield port unless port.to_i == 443
-            else
-              yield port unless port.to_i == 80
-            end
-          end
+        delegate :tld_length, :tld_length=, :secure_protocol, :secure_protocol=,
+          :extract_domain, :extract_subdomain, :extract_subdomains, :url_for,
+          :full_url_for, :path_for, to: ActionDispatch::Http::URI
```

Comment:
This is just a docs/api question, but I wonder what the purpose of having both a URL and URI class in the public domain, and if we just choose one (URL) because it's already public API than we should nodoc the URI class and keep the documentation for those public methods on the URL class (you can use `# :method: extract_domain` style rdoc notation).

Reply:
As for the public API, it will be less confusing to have one URL class. The idea is to move the rdoc comments for the delegated methods to the URL class from the URI class, and then make the URI class nodoc, right?

```suggestion
# Returns the domain part of a host given the domain level.
        #   :method: extract_domain
        # 
        #    # Top-level domain example
        #    extract_domain('www.example.com', 1) # => "example.com"
        #    # Second-level domain example
        #    extract_domain('dev.www.example.co.uk', 2) # => "example.co.uk"
        delegate :tld_length, :tld_length=, :secure_protocol, :secure_protocol=,
          :extract_domain, :extract_subdomain, :extract_subdomains, :url_for,
          :full_url_for, :path_for, to: ActionDispatch::Http::URI
```